db/compaction/compaction_iterator.cc

```diff
@@ -625,24 +625,32 @@ void CompactionIterator::NextFromInput() {
 
           TEST_SYNC_POINT_CALLBACK(
               "CompactionIterator::NextFromInput:SingleDelete:2", nullptr);
-          if (next_ikey.type == kTypeSingleDeletion ||
-              next_ikey.type == kTypeDeletion) {
+          if (next_ikey.type == kTypeSingleDeletion) {
             // We encountered two SingleDeletes for same key in a row. This
             // could be due to unexpected user input. If write-(un)prepared
             // transaction is used, this could also be due to releasing an old
             // snapshot between a Put and its matching SingleDelete.
-            // Furthermore, if write-(un)prepared transaction is rolled back
-            // after prepare, we will write a Delete to cancel a prior Put. If
-            // old snapshot is released between a later Put and its matching
-            // SingleDelete, we will end up with a Delete followed by
-            // SingleDelete.
             // Skip the first SingleDelete and let the next iteration decide
-            // how to handle the second SingleDelete or Delete.
+            // how to handle the second SingleDelete.
 
             // First SingleDelete has been skipped since we already called
             // input_.Next().
             ++iter_stats_.num_record_drop_obsolete;
             ++iter_stats_.num_single_del_mismatch;
+          } else if (next_ikey.type == kTypeDeletion) {
+            ROCKS_LOG_FATAL(
+                info_log_,
+                "Found SD and type:%d on the same key, violating the contract "
+                "of SingleDelete. Check your application to make sure the "
+                "application does not mix SingleDelete and Delete for "
+                "the same key. If you are using "
+                "write-prepared/write-unprepared transactions, and use "
+                "SingleDelete to delete certain keys, then make sure "
+                "TransactionDBOptions::rollback_deletion_type_callback is "
+                "configured properly. Mixing SD and DEL can lead to undefined "
+                "behaviors",
+                static_cast<int>(next_ikey.type));
```

Comment:
Can we return a `Status::Corruption` with the message instead of crashing the process?

Reply:
We can set `status_` and return.